dapp/src/components/buySell/ApproveSwap.js

```diff
@@ -0,0 +1,264 @@
+      !coinApproved
+    )
+    ContractStore.update((s) => {
+      s.approvalNeeded = approval
+    })
+  }, [selectedSwap, needsApproval])
+
+  const {
+    vault,
+    flipper,
+    uniV3SwapRouter,
+    uniV2Router,
+    sushiRouter,
+    curveOUSDMetaPool,
+    usdt,
+    dai,
+    usdc,
+    ousd,
+  } = useStoreState(ContractStore, (s) => s.contracts || {})
+
+  const contracts = {
+    vault: vault,
+    flipper: flipper,
+    uniswap: uniV3SwapRouter,
+    curve: curveOUSDMetaPool,
+    uniswapV2: uniV2Router,
+    sushiswap: sushiRouter,
+  }
+
+  useEffect(() => {
+    if (selectedSwap) {
+      if (
+        isApproving.contract === selectedSwap.name &&
+        isApproving.coin === stableCoinToApprove
+      ) {
+        setStage('waiting-network')
+      } else {
+        setCoinApproved(false)
+        setStage('approve')
+      }
+    }
+  }, [selectedSwap])
+
+  useEffect(() => {
+    const coinToContract = { dai, usdt, usdc, ousd }
+    if (Object.keys(coinToContract).includes(stableCoinToApprove)) {
+      setContract(coinToContract[stableCoinToApprove])
+    }
+  }, [stableCoinToApprove, usdt, dai, usdc, ousd])
+
+  const ApprovalMessage = ({
+    stage,
+    selectedSwap,
+    stableCoinToApprove,
+    isMobile,
+  }) => {
+    const capitalized = selectedSwap.name.charAt(0).toUpperCase()
+    const noncapitalized =
+      selectedSwap.name === 'uniswapV2'
+        ? selectedSwap.name.slice(1, 7)
+        : selectedSwap.name.slice(1)
+    const origin =
+      selectedSwap.name === 'flipper' || selectedSwap.name === 'vault'
+        ? 'the '
+        : ''
+    const vault = selectedSwap.name === 'vault' ? 'Origin ' : ''
+    const coin = stableCoinToApprove.toUpperCase()
+    const route = `${origin} ${vault} ${capitalized}${noncapitalized} to use your ${coin}`
+    const routeMobile = `${origin} ${vault} ${capitalized}${noncapitalized}`
+    if (stage === 'waiting-user') {
+      return fbt(
+        'Waiting for you to confirm...',
+        'Waiting for you to confirm...'
+      )
+    }
+    if (stage === 'waiting-network') {
+      const waitingNetworkMessage = `${origin} ${vault} ${capitalized}${noncapitalized}`
+      return fbt(
+        'Approving ' +
+          fbt.param('waiting-network', waitingNetworkMessage) +
+          '...',
+        'Approving contract'
+      )
+    }
+    if (stage === 'done') {
+      const doneMessage = `${vault} ${capitalized}${noncapitalized}`
+      return fbt(
+        fbt.param('approval-done', doneMessage) + ' approved',
+        'Contract approved'
+      )
+    }
+    return (
+      <>
+        {isMobile
+          ? fbt(
+              'Approve ' + fbt.param('route-mobile', routeMobile),
+              'Approve contract'
+            )
+          : fbt('Allow ' + fbt.param('route', route), 'Approve contract')}
+      </>
+    )
+  }
+
+  return (
+    <>
+      <button
+        className={`btn-blue buy-button mt-4 mt-md-3 w-100`}
+        hidden={
+          (!selectedSwap ||
```

Comment:
Seems very similar code to lines ~33 in this file. Could this be represented as an `approvalNeeded` bool and used in both places?